src/utils/analytics.utils.ts

```diff
@@ -0,0 +1,117 @@
+import { EntityManager } from 'typeorm';
+import {
+  addMonths,
+  addDays,
+  startOfYear,
+  subYears,
+  format,
+  addYears,
+} from 'date-fns';
+import { EventStatus, Frequency, Months } from 'src/common/constants';
+import { User } from 'src/github/entities/user.entity';
+import { FilterDto } from 'src/analytics/dto/filter.dto';
+
+export const buildQueryBuilder = (
+  entityManager: EntityManager,
+  userId: string,
+  filter: FilterDto,
+) => {
+  const queryBuilder = entityManager
+    .getRepository(User)
+    .createQueryBuilder('user')
+    .leftJoinAndSelect('user.repo', 'repo')
+    .leftJoinAndSelect('repo.pullRequests', 'pullRequest')
+    .where('user.githubUserId = :githubUserId', { githubUserId: userId });
+
+  if (filter?.repoName) {
+    queryBuilder.andWhere('repo.name = :name', { name: filter.repoName });
+  }
+
+  return queryBuilder;
```

Comment:
PullProofAI

The `buildQueryBuilder` function seems well-structured and handles building the query builder based on the provided parameters. Consider adding error handling in case of any exceptions during query building.
```diff
+   if (!entityManager || !userId || !filter) {
+       throw new Error('Invalid parameters provided for building query builder');
+   }
```



<!-- This is an auto-generated comment by PullProofAI -->